source/extensions/filters/network/ratelimit/ratelimit.cc

```diff
@@ -28,6 +33,32 @@ Config::Config(const envoy::extensions::filters::network::ratelimit::v3::RateLim
   }
 }
 
+std::vector<RateLimit::Descriptor>
+Config::applySubstitutionFormatter(std::vector<RateLimit::Descriptor> original_descriptors,
+                                   StreamInfo::StreamInfo& stream_info) {
+
+  std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+  for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+    RateLimit::Descriptor new_descriptor;
+    for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+      std::string value = descriptorEntry.value_;
+      value = formatValue(value, stream_info);
+      new_descriptor.entries_.push_back({descriptorEntry.key_, value});
+    }
+    dynamicDescriptors.push_back(new_descriptor);
+  }
+  return dynamicDescriptors;
+}
+
+std::string Config::formatValue(std::string descriptor_value, StreamInfo::StreamInfo& stream_info) {
+
+  Formatter::FormatterImpl formatter(descriptor_value);
```

Comment:
Again, the idea is to pre-create these at config load time. Then you store a vector of these, and use them during apply.

Reply:
I'm fine with all of the other changes but I think that the performance hit for creating the Formatter and cleaning it out each substitution is not noticeable relative to having to keep a vector filled with formatter objects for especially complex and verbose rate limiting schemes. That being said I'm not as familiar with this code base as you are so I will defer. 

Making the other changes rn.

Reply:
@mattklein123 the FormatterImpl class doesn't have the copy functionality/relevant interfaces needed to be used with vectors it looks like. Should we change the FormatterImpl?

Reply:
No, you can store them in a pointer if needed.

/wait